crates/typst/src/model/block.rs

```diff
@@ -0,0 +1,100 @@
+use std::any::Any;
+use std::fmt::{self, Debug};
+use std::hash::{Hash, Hasher};
+
+use smallbox::{smallbox, SmallBox};
+
+/// A small block storage for storing stylechain values
+/// either on the stack (if they fit) or on the heap.
+pub struct Block(SmallBox<dyn Blocked, S3>);
+
+impl Block {
+    /// Creates a new block.
+    pub fn new<T: Blockable>(value: T) -> Self {
+        Self(smallbox!(value))
+    }
+
+    /// Downcasts the block to the specified type.
+    pub fn downcast<T: Blockable>(&self) -> Option<&T> {
+        self.0.as_any().downcast_ref()
+    }
+
+    /// Downcasts mutably the block to the specified type.
+    pub fn downcast_mut<T: Blockable>(&mut self) -> Option<&mut T> {
+        self.0.as_any_mut().downcast_mut()
+    }
+}
+
+impl Hash for Block {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.dyn_hash(state);
+    }
+}
+
+impl Clone for Block {
+    fn clone(&self) -> Self {
+        Self(self.0.dyn_clone())
+    }
+}
+
+impl Debug for Block {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.dyn_debug(f)
+    }
+}
+
+/// A value that can be stored in a block.
+///
+/// Auto derived for all types that implement [`Any`], [`Clone`], [`Hash`],
+/// [`Debug`], [`Send`] and [`Sync`].
+pub trait Blockable: Any + Clone + Hash + Debug + Send + Sync {}
```

Comment:
I think we can get rid of this extra trait if we set things up like for `Dynamic` in `value.rs`.